### ISSUE TITLE
[BUG] Dark mode cannot be set per user

### ISSUE THREAD

**Issue by sshock** [opensearch-project/OpenSearch-Dashboards#2159]
**Describe the bug**

Dark Mode should be a user-configurable setting.

**To Reproduce**
1. Go into Stack Management => Advanced Settings
2. Turn dark mode on or off

**Expected behavior**
Dark mode is turned on/off only for users who wish to inherit the global default.

**Actual behavior**
Dark mode is turned on/off for everyone in the entire large company and users have no way to override it.

**OpenSearch Version**
1.3.2

**Dashboards Version**
1.3.2

**Comment by peternied** [comment]
@kavilla This is a good feature request, but the security team isn't focused on user experience changes of this nature.  If there is a data store needed that handles per user configuration, it might make sense for that to be implemented as part of the security plugin dashboards codebase, but the features and functionality that are offered using that store should be part of Dashboards.

**Comment by kavilla** [comment]
> @kavilla This is a good feature request, but the security team isn't focused on user experience changes of this nature. If there is a data store needed that handles per user configuration, it might make sense for that to be implemented as part of the security plugin dashboards codebase, but the features and functionality that are offered using that store should be part of Dashboards.

@peternied,

It's a little bit of a gray area. The concept of user and tenancy does not currently exist for OSD. The problem occurs how the security dashboards introduces tenants and how the data is stored with a per user configuration. 

Due multitenancy creating multiple system indices, each index has it's own document with a  type of `config`. This `config` document has the settings values for dark mode. From my understanding what @sshock is requesting that private tenants should be using the `config` document by a system admin for settings instead of each private tenant having their own document to pull configs from. Which is basically, not compatible with the current state of the security plugin and there's nothing we could do from our side from a UX perspective.

>  For UX features, please transfer this back to the Dashboards repository.
> This bug should be where the team invested in resolving this issue can support the work.

@cwperks

This seems surprising to me and I was not made aware. We have multiple OpenSearch Dashboards external plugins that we do not maintain from a core perspective. Doesn't seem to make sense to me that if there is a bug in an external plugin that the bug should exist in the core repo and the core repo's maintainers are responsible for fixing it. We also do not have the proper expertise as a maintainer of the security dashboards team would have.

could you provide some clarity there? Is it due to specifically with the security plugin and the planned work to roll permissions in OpenSearch Dashboards?

**Comment by peternied** [comment]
@kavilla I think we've found a troublesome area in our breakdown of features within OpenSearch and Dashboards, I think we should have a discussion to iron out our expectations and how we want to resolve this issue - could you set something up?

**Comment by sshock** [comment]
Also, I think there may still be some confusion.  @kavilla We're not even using private tenants at my company, and my goal is not at all for private tenants to be able to inherit settings from the Global tenant.

If anything it is quite the opposite of that.  It is for users to be able to choose their own setting for dark mode, which I assume they already can within their own private tenant and that won't affect anyone else, but the problem is users can't set it within the Global tenant (or a custom tenant) without affecting everyone else.

While using the "Global" tenant there is only a single setting for Dark mode that all users in the entire company get forced on them with no way to override.  One choice affects all the hundreds or thousands of employees and they all have to like it and live with it.  Or even worse, if everyone is allowed to change settings in the global tenant, now you have an even crazier situation, which is all day long every day having it change back and forth between light and dark mode at the whim of each random user who goes in and changes it, affecting everyone else multiple times throughout the day.

I don't know how user settings work or if user settings are even an actual thing right now, though I imagine it must be a real thing because somewhere it keeps track of what my roles are and which tenant I am currently on.  Wherever it keeps track of "what is my current tenant", maybe that is where it could also keep track of "do I want dark mode or not".

**Comment by peternied** [comment]
@sshock First off, thanks for bearing with us as we get our house in order.  With our implementation, we are not planning on support this scenario.

However, for the initiatives of Dashboards - see [issue #1405](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/1405), `Dashboards Anywhere` will enable features like customized settings per user.

@kavilla How would you like to resolve this issue? - I think the next step is to close it, but you might have a way to map it into the future plans.

**Comment by kgcreative** [comment]
@peternied & @kavilla -- perhaps we tag this as a feature request instead of a bug? That will allow us to keep it open, and close it when we do implement the changes resolve this.

**Comment by sshock** [comment]
I'm definitely in favor of calling it a feature instead of a bug if that is needed to keep it open.  I'll change the title but will need someone else to change the tag.

Any idea on the timeline of the `Dashboards Anywhere` initiative?

Finally, is there any way to work around this until then?  For example, can an individual force light or dark mode using a browser extension like Stylus?

Thanks!

**Comment by joshuarrrr** [comment]
I just wanted to highlight what I think is a key takeaway from @sshock:

> I don't know how user settings work or if user settings are even an actual thing right now, though I imagine it must be a real thing because somewhere it keeps track of what my roles are and which tenant I am currently on. Wherever it keeps track of "what is my current tenant", maybe that is where it could also keep track of "do I want dark mode or not".

It seems that our current data model overloads the system index with a bunch of heterogeneous settings. Within the core dashboards repo, this approach, while not exactly clean, causes no problems because "user" is not really a core concept. But for plugins that do need to distinguish between different types of settings (such as the `security` plugin), I think it's reasonable to expect them to be better separated by usage, rather than use the system index for any data that needs persistence.

**Comment by cliu123** [comment]
@sshock Yes, with the current design,the Advanced Settings is tied with tenants. The logic behind the design is that tenant intends to provide consistent user experience for all users selecting a same tenant. But your request is actually valid, it would help if we can have flexibility for users to configure personal preferences that can be applied to all tenants globally. Thanks for bringing this up!

**Comment by ThiefMaster** [comment]
Can't you use the client's preference instead of making this a server-side change?

https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme

**Comment by kgcreative** [comment]
> Can't you use the client's preference instead of making this a server-side change?
> 
> https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme

In the fullness of time, yes. We will add a "Dark | Light | Match System Preferences" -- however, our CSS stack is woefully out of date, and that's not something that the current theming system can support, so it requires some additional work to refactor OUI to accomodate for this

**Comment by flytrader** [comment]
> **Describe the bug**
> 
> Dark Mode should be a user-configurable setting.
> 
> **To Reproduce** Steps to reproduce the behavior:
> 
> 1. Go into Stack Management => Advanced Settings
> 2. Turn dark mode on or off
> 
> **Expected behavior** Dark mode is turned on/off only for users who wish to inherit the global default.
> 
> **Actual behavior** Dark mode is turned on/off for everyone in the entire large company and users have no way to override it.
> 
> **OpenSearch Version** 1.3.2
> 
> **Dashboards Version** 1.3.2

It can be modified per user. All the NonDefault values that are changed in a users Private Tenant/Advanced settings is visible with the GET command. You can also see if they did change something and reverted it to default (see last example).

curl -X GET -u "user:password" "opensearchendpoint/.kibana_<USER>/_search"

{
                "_index": ".kibana_-1224620319_test12345_1",
                "_id": "config:2.11.0",
                "_score": 1,
                "_source": {
                    "config": {
                        "buildNum": 301202401,
                        "defaultIndex": "indexnumber",
                        "theme:darkMode": true,
                        "dateFormat:dow": "Monday",
                        "history:limit": 666
                    },
                    "type": "config",
                    "references": [],
                    "migrationVersion": {
                        "config": "7.9.0"
                    },
                    "updated_at": "2024-06-14T13:09:07.029Z"
                }
                
With Default values exept the theme:darkMode
                
                 "config": {
                        "buildNum": 301202401,
                        "defaultIndex": "indexnumber",
                        "theme:darkMode": true,
                        "dateFormat:dow": null,
                        "history:limit": null
                    }

**Comment by mthemis-provenir** [comment]
Has there been any progress on this since the discussions last year? We're facing a problem here whereby most people prefer dark mode, but for others it makes it hard to read due to how their eyes work. I'd consider this not a regular feature, but an accessibility feature.

It can be worked around by the users setting up their private tenant with the various index patterns, but it means configuration drift when we add / remove / update them in the global tenant.

**Comment by svichkar** [comment]
?

**Comment by amirkkn** [comment]
I think the issue can be closed as we have this feature in v3.2.0

**Comment by andy-townsend** [comment]
> I think the issue can be closed as we have this feature in v3.2.0

I don't see anything in the release notes to support this, it still doesn't seem to be supported on a per-user basis.
